### PR TITLE
Local adjustments - improvments of GUI - improve Wavelet pyramid

### PR DESCRIPTION
**Pull request by Desmis**
- For each spot, users can now choose which local tool he wants to add or remove;
 - pp3 files are now slimmed down: for each spot, only associated local tool parameters are saved;
 - For copy/paste or save/load, using Partial Dialog, users can now select spots they want to keep;
 - Optimisation: When switching between spot, picture is no longer updated if not necessary;
 - Locallab is not anymore possible in batch mode (quite useless to maintain this code for local adjustments...);

Thanks to @pandagrapher

For this to work properly, you need to **"clear" the cache**.
Admittedly this introduces a loss of compatibility with the old pp3, for those who used "newlocallab", but it is difficult to do otherwise.

There are also, many improvments to wavelet pyramid and fixed numerous bugs

### CODE REVIEW

File: rtengine/procparams.cc [hunk 1]
```diff
@@ -390,51 +400,51 @@ bool ToneCurveParams::operator !=(const ToneCurveParams& other) const
 RetinexParams::RetinexParams() :
     enabled(false),
     cdcurve{
-        DCT_Linear
```
Comment (Floessie) [inline rtengine/procparams.cc]: @Desmis @Pandagrapher Obviously `astyle` destroyed the formatting. Would you mind to repair it? Afterwards it's easier to review this file.

Comment (Desmis) [reply rtengine/procparams.cc]: @Floessie 
I just push a "small" change tp procparams.cc

**Comment by Floessie** [comment]
Builds fine with GCC 9 and 10 on Debian Testing AMD64.

General remark: "adjustement" -> "adjustment".

A lot of warnings with Clang 10:

[newlocallab_clang10_warnings.txt](https://github.com/Beep6581/RawTherapee/files/4527800/newlocallab_clang10_warnings.txt)

A warning from GCC on non-SSE builds (i386, PPC32):

```
In file included from /home/user/src/rawtherapee/rtengine/curves.h:30,
                 from /home/user/src/rawtherapee/rtengine/iplocallab.cc:28:
/home/user/src/rawtherapee/rtengine/sleef.h: In function ‘_ZN8rtengine15ImProcFunctions11maskcalccolEibbiiiiiiiPNS_8LabImageES2_S2_S2_S2_iRNS_12local_paramsEfbRKNS_14LocCCmaskCurveERbRKNS_14LocLLmaskCurveES8_RKNS_14LocHHmaskCurveES8_SE_S8_bbbbbbbffffffiffR3LUTIfES8_RKNS_11LocwavCurveES8_iiiiibfffffffffi._omp_fn.36’:
/home/user/src/rawtherapee/rtengine/sleef.h:933:65: warning: ‘newhr’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 __inline float mlaf(float x, float y, float z) { return x * y + z; }
                                                                 ^
/home/user/src/rawtherapee/rtengine/iplocallab.cc:4294:27: note: ‘newhr’ was declared here
                     float newhr;
                           ^~~~~
```

HTH,
Flössie

**Comment by Pandagrapher** [comment]
@Desmis 
As I cannot directly commit in newlocallab branch, gonna send you a patch for the warnings in GUI functions

Pierre

**Comment by Desmis** [comment]
@Floessie 
I fixed the warning in ipretinex.cc iplocallab.cc


But why @Pandagrapher  cannot directly push a change ???

**Comment by Desmis** [comment]
@Pandagrapher 
The simplest thing would not be that you modify your version of newlocallab, then you make a pull-request, then I make a pull request, if it is possible , and if it works ?

@Beep6581 
it is incomprehensible, Pierre makes very important modifications and he has no commit rights
How can we quickly resolve this. 

jacques

**Comment by Desmis** [comment]
@Floessie 

All warnings are now fixed (I hope), either by me, either by @Pandagrapher 

Jacques

**Comment by Floessie** [comment]
> it is incomprehensible, Pierre makes very important modifications and he has no commit rights
How can we quickly resolve this.

I'm with you. IMHO it would be great if Pierre got commit rights. :+1: Kindly pinging @Beep6581.

**Comment by Desmis** [comment]
@Floessie   @heckflosse  
Have you an idea how to fix warning "lgtm" above
It seems to be reported to "labimage.cc" -> copyfrom

Thank you

jacques

**Comment by Benitoite** [comment]
Having a real number scaler in the fixed line converted you to a default minimum 64b long long in c++11 so no potential overflow warning there.  Scaling by `* 1` might fool the analyzer but be optimized out by the compiler. What do you think?
A8F44306-2C5C-42E6-80CE-ECB459E19DBA

**Comment by Desmis** [comment]
@Benitoite 
I just tried your patch...same result !

Thank you :)

**Comment by Benitoite** [comment]
Thanks for trying.  LGTM must optimize it out as well.  As I understand it multiplying by 1 causes 11 instructions.

**Comment by Desmis** [comment]
@Floessie 
Thank you for these improvments :)

jacques

**Comment by Desmis** [comment]
@Floessie 

I found a big **bug** with the last commit 48ea474 **restore procparams.cc** formatting

With the changes, now 
* when you use "local adjustments"
* open one tool
* enable this tool (for example local contrast and wavelet)
* close the windows preview.... the system **crash**

I tried with 
git checkout 4fc3278   (the last commit for labimage.cc), all works well

I am looking to procparams.cc  :)
jacques

**Comment by Floessie** [comment]
@Desmis Sorry. Looking...

**Comment by Floessie** [comment]
@Desmis Can't make it crash, although I don't really understand point 4). Is moving to another image (F3/F4) enough?

**Comment by Floessie** [comment]
@Desmis I think I found it...

**Comment by Desmis** [comment]
@Floessie 

Excuse my bad english

open RT
clean cache (to be sure there is no other factors)
choose one file (one image)
Open it
Go to local adjustemnts
add one RT-spot
add one tool
close the window

System crash, at lesat with my computer :)

**Comment by Floessie** [comment]
Now that I can reproduce it, the backtrace points exactly to the place I suspected. Pushing soon.

**Comment by Floessie** [comment]
@Desmis Can you verify the fix?

**Comment by Desmis** [comment]
@Floessie 

All seems to work now, thank you :)

Please do not psuh a commit this afternoon, I am working on "wavelet pyramid residual"

Thank you

Jacques

**Comment by Desmis** [comment]
"Local adjustments" advancement,  including the GUI part performed by @Pandagrapher , lets "Merge pull request"

https://discuss.pixls.us/t/local-adjustments-improvment-of-gui-improve-wavelet-pyramid/17741/8

if nobody objects, i will do this "Merge Pull Request" this weekend.

jacques

**Comment by Desmis** [comment]
@heckflosse 

Ingo
You can merge with https://discuss.pixls.us/t/sometimes-1-pixel-artifacts-from-x-trans-demosaicing/18179/2

But, I have a mistake, in d1b8c2b then corrected, and I have always LGTM not successfull
Could you have a look

I don't know how to go back...

If you want, you could ignore all this last work...
You merge with dev
Then I re-commit with the changes (in a good way!)

Thank you
Jacques

**Comment by Desmis** [comment]
@heckflosse 
Hello Ingo

I have create a branch newlocallab2 which stop at 24bad48

I think, to verify, you can use this branch... 

Jacques

**Comment by Desmis** [comment]
@heckflosse 
Ingo

In case of, I rebuild with newlocallab2
Merge with dev
Then my last bad changes

jacques

**Comment by Desmis** [comment]
I just create a new PULL REQUEST with newlocallab2, always in case of LGTM fails

**Pull request closed without merging**